File: day2.hs
import Data.List.Split

main = do
    let file = "input2.txt"
    contents <- fmap lines (readFile file)
    print $ length $ isValid contents
    print $ length $ isValid2 contents

data Password = Password {
    min:: String,
    max:: String,
    letter:: Char,
    value:: String
}

parsePassword :: String -> Password
parsePassword p = Password mn mx l v
        where letters = words p
              ([mn, mx]) = splitOn ("-") (letters !! 0)
              l = head (letters !! 1)
              v = letters !! 2

filterPswd :: Password -> Bool
filterPswd (Password a b c d) = count >= (read a) && count <= (read b)
    where count = length $ filter(==c) d

isValid :: [String] -> [Password]
isValid passwords = filter filterPswd $ map parsePassword passwords

isValid2 :: [String] -> [Password]
isValid2 passwords = filter filterPswd2 $ map parsePassword passwords

filterPswd2 :: Password -> Bool
filterPswd2 (Password a b c d) = (d !! ((read a) - 1) == c) /= (d !! ((read b) -1) == c)

File: 2020/day2/day2.js
fs = require("fs");

lines = fs.readFileSync("./input2.txt", "utf8");
data = lines.split("\n").filter(String);

function splitString(str) {
  let [x, y, value] = str.split(" ");
  let [min, max] = x.split("-");
  let letter = y.replace(":", "");
  return [min, max, letter, value];
}

//solution 1
function isCorrect(str) {
  [min, max, letter, value] = splitString(str);
  totalLetters = 0;
  for (i = 0; i < value.length; i++) {
    if (value[i] == letter) {
      totalLetters++;
    }
  }
  return totalLetters >= min && totalLetters <= max;
}

console.log(data.filter(isCorrect).length);

// solution 2
function isCorrect2(str) {
  [min, max, letter, value] = splitString(str);
  charAtMin = value.charAt(Number(min) - 1);
  charAtMax = value.charAt(Number(max) - 1);

  return (charAtMin == letter) != (charAtMax == letter);
}

console.log(data.filter(isCorrect2).length);
